plugins/inventory/digitalocean.py

```diff
@@ -0,0 +1,238 @@
+    required: true
+    choices: ['community.digitalocean.digitalocean']
+  api_token:
+    description:
+     - DigitalOcean OAuth token.
+    required: true
+    type: str
+    aliases: [ oauth_token ]
+    env:
+      - name: DO_API_TOKEN
+  attributes:
+    description: >-
+      Droplet attributes to add as host vars to each inventory host.
+      Check out the DO API docs for full list of attributes at
+      U(https://developers.digitalocean.com/documentation/v2/#list-all-droplets).
+    type: list
+    elements: str
+    default:
+      - id
+      - name
+      - networks
+      - region
+      - size_slug
+  var_prefix:
+    description:
+      - Prefix of generated varible names (e.g. C(tags) -> C(do_tags))
+      - Has to start with a letter,
+        the only valid special character is underscore (C(_)).
+    type: str
+    default: 'do_'
+  pagination:
+    description:
+      - Maximum droplet objects per response page.
+      - If the number of droplets related to the account exceeds this value,
+        the query will be broken to multiple requests (pages).
+      - This value cannot be greater than 200.
+    type: int
+    default: 200
+'''
+
+EXAMPLES = r'''
+# Using keyed groups and compose for hostvars
+plugin: community.digitalocean.digitalocean
+api_token: "{{ api_token }}"
+attributes:
+  - id
+  - name
+  - memory
+  - vcpus
+  - disk
+  - size
+  - image
+  - networks
+  - volume_ids
+  - tags
+  - region
+keyed_groups:
+  - key: do_region.slug
+    prefix: 'region'
+    separator: '_'
+  - key: do_tags | lower
+    prefix: ''
+    separator: ''
+compose:
+  ansible_host: do_networks.v4 | selectattr('type','eq','public')
+    | map(attribute='ip_address') | first
+  class: do_size.description | lower
+  distro: do_image.distribution | lower
+'''
+
+import re
+import json
+from ansible.errors import AnsibleParserError
+from ansible.module_utils.urls import Request
+from ansible.module_utils.six.moves.urllib.error import URLError, HTTPError
+from ansible.plugins.inventory import BaseInventoryPlugin, Constructable, Cacheable
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):
+
+    NAME = 'community.digitalocean.digitalocean'
+
+    def _validate_config(self, config):
+        if config['plugin'] != 'community.digitalocean.digitalocean':
+            raise AnsibleParserError("plugin doesn't match this plugin")
+        do_fields = ['id', 'name', 'memory', 'vcpus', 'disk', 'locked', 'created_at', 'status', 'backup_ids', 'snapshot_ids', 'features',
+                     'region', 'image', 'size', 'size_slug', 'networks', 'kernel', 'next_backup_window', 'tags', 'volume_ids', 'vpc_uuid']
+
+        try:
+            attributes = config['attributes']
+            for a in attributes:
+                if a not in do_fields:
+                    raise AnsibleParserError("invalid attribute: " + a)
+        except KeyError:
+            pass
```

Comment:
I agree it's probably best to not validate this (and same for `pagination` below). This will be a constant maintenance burden if anything ever changes on DO's side.

Validating against an actual response would be possible, but that would make everything more complex (also there's the case when the response contains no entries - do we skip validation then?).

Reply:
I had that feeling too, but wanted to try all possibilities I can. Let's throw it out!